crates/query-engine/translation/src/translation/query/fields.rs

```diff
@@ -318,3 +323,201 @@ fn translate_nested_field(
         },
     ))
 }
+
+#[allow(clippy::too_many_arguments)]
+/// In order to return the expected type representation for each column,
+/// we need to wrap columns in type representation cast, and unpack composite types
+/// so we can wrap them.
+fn unpack_and_wrap_fields(
+    env: &Env,
+    state: &mut State,
+    current_table: &TableNameAndReference,
+    join_relationship_fields: &mut Vec<relationships::JoinFieldInfo>,
+    column: &str,
+    alias: sql::ast::ColumnAlias,
+    type_info: &CompositeTypeInfo<'_>,
+    nested_field_joins: &mut Vec<JoinNestedFieldInfo>,
+) -> Result<(sql::ast::ColumnAlias, sql::ast::Expression), Error> {
+    let column_info = type_info.lookup_column(column)?;
+
+    // Different kinds of types have different strategy for converting to their
+    // type representation.
+    match column_info.r#type {
+        // Scalar types can just be wrapped in a cast.
+        Type::ScalarType(scalar_type) => {
+            let column_type_representation = env.lookup_type_representation(&scalar_type);
+            let (alias, expression) = sql::helpers::make_column(
+                current_table.reference.clone(),
+                column_info.name.clone(),
+                alias,
+            );
+            Ok((
+                alias,
+                wrap_in_type_representation(expression, column_type_representation),
+            ))
+        }
+        // Composite types are a more involved case because we cannot just "cast"
+        // a composite type, we need to unpack it and cast the individual fields.
+        // In this case, we unpack a single composite column selection to an explicit
+        // selection of all fields.
+        Type::CompositeType(ref composite_type) => {
+            // build a nested field selection of all fields.
+            let nested_field = unpack_composite_type(env, composite_type)?;
+
+            // translate this as if it is a nested field selection.
+            let (nested_field_join, nested_column_reference) = translate_nested_field(
+                env,
+                state,
+                current_table,
+                &column_info,
+                nested_field,
+                join_relationship_fields,
+            )?;
+
+            nested_field_joins.push(nested_field_join);
+
+            Ok((
+                alias,
+                sql::ast::Expression::ColumnReference(nested_column_reference),
+            ))
+        }
+        // TODO: Arrays of composite types are not handled yet.
```

Comment:
I am confused; it looks like they are. Should this say "Arrays of arrays are not handled yet"?

Reply:
Thanks! stale comment.